src/main.rs

```diff
@@ -41,44 +35,105 @@ fn main() -> Result<()> {
+                                    devices::create(&url, id, data, app_id)?
+                                },
+                            }
+                        }
+                        _ => println!("{}", NO_COMMAND_MESSAGE)
+
+                    }
+
+                }
+                Verbs::delete => {
+
+                    match cmd.subcommand() {
+
+                        (res, Some(command)) => {
+                            let id = command.value_of(Parameters::id).unwrap();
+                            let nam = Resources::from_str(res);
+                
+                            match nam? {
+                                Resources::app => apps::delete(&url, id)?,
+                                Resources::device => {
+                                    let app_id = command.value_of(Resources::app).unwrap();
+                                    devices::delete(&url, app_id, id)?
+                                },
+                            }
+                        }
+                        _ => println!("{}", NO_COMMAND_MESSAGE)
+
+                    }
+
+                }
+                Verbs::edit => {
+
+                    match cmd.subcommand() {
+
+                        (res, Some(command)) => {
+                            let id = command.value_of(Parameters::id).unwrap();
+                            
+                            let nam = Resources::from_str(res);
+                
+                            match nam? {
+                                Resources::app => apps::edit(&url, id),
+                                Resources::device => {
+                                    let app_id = command.value_of(Resources::app).unwrap();
+                                    devices::edit(&url, app_id, id)
+                                },
+                            }
+                        }
+                        _ => println!("{}", NO_COMMAND_MESSAGE)
+
+                    }
+
+
+                }
+                Verbs::get => {
+
+                    match cmd.subcommand() {
+
+                        (res, Some(command)) => {
+                            let id = command.value_of(Parameters::id).unwrap();
+                            
+                            let nam = Resources::from_str(res);
+                
+                            match nam? {
+                                Resources::app => apps::read(&url, id)?,
+                                Resources::device => {
+                                    let app_id = command.value_of(Resources::app).unwrap();
+                                    devices::read(&url, app_id, id)?
+                                },
+                            }
+                        }
+                        _ => println!("{}", NO_COMMAND_MESSAGE)
+                    }
+
+                }
             }
+
         }
+        _ => println!("No command supplied. Run `drgctl help` to see available commands.")
```

Comment:
I would simply use the clap.rs flag to force args : https://docs.rs/clap/2.31.1/clap/enum.AppSettings.html#variant.ArgRequiredElseHelp

Reply:
that's what `unreachable!()` is for :)

Reply:
You are right, but this match arm not required anymore as clap can handle it with `.setting(AppSettings::SubcommandRequiredElseHelp)`